File: src/Web/Client/Services/TasksService.cs
using Flurl;
using MudBlazor;
using Shared.Dtos;
using Shared.ViewModels;

namespace Web.Client.Services;

public class TasksService : ApiService
{
    public TasksService(HttpClient httpClient, ISnackbar snackbar) : base(httpClient, snackbar)
    {
    }

    public async Task<TasksVM?> GetAll(Guid projectId)
        => await Get<TasksVM>("tasks", ProjectIdHeader(projectId));

    public async Task<TaskVM?> Get(Guid projectId, int shortId)
        => (await Get<TasksVM>($"tasks/{shortId}", ProjectIdHeader(projectId)))?.Tasks.Single();

    public async Task<TaskVM?> Get(Guid projectId, IEnumerable<Guid> ids)
    {
        var url = "tasks"
            .SetQueryParam("ids", ids)
            .ToString();
        return (await Get<TasksVM>(url, ProjectIdHeader(projectId)))?.Tasks.Single();
    }

    public async Task<bool> Create(Guid projectId, CreateTaskDto model)
        => await Post("tasks", model, ProjectIdHeader(projectId));

    public async Task<TaskCommentsVM?> GetComments(Guid projectId, Guid taskId)
        => await Get<TaskCommentsVM>($"tasks/{taskId}/comments", ProjectIdHeader(projectId));

    public async Task<TaskActivitiesVM?> GetActivities(Guid projectId, Guid taskId)
        => await Get<TaskActivitiesVM>($"tasks/{taskId}/activities", ProjectIdHeader(projectId));

    public async Task<bool> UpdateStatus(Guid projectId, Guid taskId, UpdateTaskStatusDto model)
        => await Post($"tasks/{taskId}/update-status", model,  ProjectIdHeader(projectId));

    public async Task<bool> UpdatePriority(Guid projectId, Guid taskId, UpdateTaskPriorityDto model)
        => await Post($"tasks/{taskId}/update-priority", model, ProjectIdHeader(projectId));

    public async Task<bool> UpdateAssignee(Guid projectId, Guid taskId, UpdateTaskAssigneeDto model)
        => await Post($"tasks/{taskId}/update-assignee", model, ProjectIdHeader(projectId));

    public async Task<bool> UpdateDescription(Guid projectId, Guid taskId, UpdateTaskDescriptionDto model)
        => await Post($"tasks/{taskId}/update-description", model, ProjectIdHeader(projectId));

    public async Task<bool> AddComment(Guid projectId, Guid taskId, AddTaskCommentDto model)
        => await Post($"tasks/{taskId}/comments", model, ProjectIdHeader(projectId));

    public async Task<bool> LogTime(Guid projectId, Guid taskId, LogTaskTimeDto model)
        => await Post($"tasks/{taskId}/log_time", model, ProjectIdHeader(projectId));

    public async Task<bool> UpdateEstimatedTime(Guid projectId, Guid taskId, UpdateTaskEstimatedTimeDto model)
        => await Post($"tasks/{taskId}/update-estimated-time", model, ProjectIdHeader(projectId));

    public async Task<TaskRelationshipsVM?> GetRelationships(Guid projectId, Guid taskId)
        => await Get<TaskRelationshipsVM>($"tasks/{taskId}/relationships", ProjectIdHeader(projectId));

    private static Headers ProjectIdHeader(Guid projectId)
        => Headers.From(("ProjectId", projectId.ToString()));
}

File: src/Web/Client/composables/tasksService.ts
import type { AddTaskLoggedTimeDto, AddTaskCommentDto, CreateTaskDto, UpdateTaskEstimatedTimeDto, UpdateTaskAssigneeDto, UpdateTaskDescriptionDto, UpdateTaskPriorityDto, UpdateTaskStatusDto, UpdateTaskTitleDto, UpdateTaskBoardDto } from "~/types/dtos/tasks";
import type { TaskRelationshipsVM, TaskActivitiesVM, TaskCommentsVM, TasksVM } from "~/types/viewModels/tasks";

export const useTasksService = () => {
    const api = useApi();

    return {
        async getTasks(projectId: string) {
            return await api.sendGetRequest<TasksVM>('tasks', { 'ProjectId': projectId });
        },
        async createTask(projectId: string, model: CreateTaskDto) {
            await api.sendPostRequest('tasks', model, { 'ProjectId': projectId });
        },
        async getTask(shortId: number, projectId: string) {
            return (await api.sendGetRequest<TasksVM>(`tasks/${shortId}`, { 'ProjectId': projectId }))?.tasks[0];
        },
        async updateDescription(id: string, projectId: string, model: UpdateTaskDescriptionDto) {
            await api.sendPostRequest(`tasks/${id}/description`, model, { 'ProjectId': projectId });
        },
        async updatePriority(id: string, projectId: string, model: UpdateTaskPriorityDto) {
            await api.sendPostRequest(`tasks/${id}/priority`, model, { 'ProjectId': projectId });
        },
        async updateAssignee(id: string, projectId: string, model: UpdateTaskAssigneeDto) {
            await api.sendPostRequest(`tasks/${id}/assignee`, model, { 'ProjectId': projectId });
        },
        async updateStatus(id: string, projectId: string, model: UpdateTaskStatusDto) {
            await api.sendPostRequest(`tasks/${id}/status`, model, { 'ProjectId': projectId });
        },
        async addComment(id: string, projectId: string, model: AddTaskCommentDto) {
            await api.sendPostRequest(`tasks/${id}/comments`, model, { 'ProjectId': projectId });
        },
        async addLoggedTime(id: string, projectId: string, model: AddTaskLoggedTimeDto) {
            await api.sendPostRequest(`tasks/${id}/logged-time`, model, { 'ProjectId': projectId });
        },
        async updateEstimatedTime(id: string, projectId: string, model: UpdateTaskEstimatedTimeDto) {
            await api.sendPostRequest(`tasks/${id}/estimated-time`, model, { 'ProjectId': projectId });
        },
        async getComments(id: string, projectId: string) {
            return await api.sendGetRequest<TaskCommentsVM>(`tasks/${id}/comments`, { 'ProjectId': projectId });
        },
        async getActivities(id: string, projectId: string) {
            return await api.sendGetRequest<TaskActivitiesVM>(`tasks/${id}/activities`, { 'ProjectId': projectId });
        },
        async updateTitle(id: string, projectId: string, model: UpdateTaskTitleDto) {
            await api.sendPostRequest(`tasks/${id}/title`, model, { 'ProjectId': projectId });
        },
        async deleteTask(id: string, projectId: string) {
            await api.sendPostRequest(`tasks/${id}/delete`, undefined, { 'ProjectId': projectId });
        },
        async updateBoard(projectId: string, model: UpdateTaskBoardDto) {
            await api.sendPostRequest('tasks/update-board', model, { 'ProjectId': projectId });
        },
        async getRelationships(id: string, projectId: string) {
            return await api.sendGetRequest<TaskRelationshipsVM>(`tasks/${id}/relationships`, { 'ProjectId': projectId });
        }
    }
}